testutils/clustermanager/gke_test.go

```diff
@@ -90,6 +92,234 @@ func (fgsc *FakeGKESDKClient) get(project, location, cluster string) (*container
+				},
+			},
+		}, {
+			// Override other parts but not zone
+			&numNodesOverride, &nodeTypeOverride, &regionOverride, nil, nil, "", "",
+			&GKECluster{
+				Request: &GKERequest{
+					NumNodes:      2,
+					NodeType:      "foonode",
+					Region:        "fooregion",
+					Zone:          "",
+					BackupRegions: []string{"us-west1", "us-east1"},
+				},
+			},
+		}, {
+			// Set env Region
+			nil, nil, nil, nil, nil, "customregion", "",
+			&GKECluster{
+				Request: &GKERequest{
+					NumNodes:      1,
+					NodeType:      "n1-standard-4",
+					Region:        "customregion",
+					Zone:          "",
+					BackupRegions: []string{"us-west1", "us-east1"},
+				},
+			},
+		}, {
+			// Set env backupzone
+			nil, nil, nil, nil, nil, "", "backupregion1 backupregion2",
+			&GKECluster{
+				Request: &GKERequest{
+					NumNodes:      1,
+					NodeType:      "n1-standard-4",
+					Region:        "us-central1",
+					Zone:          "",
+					BackupRegions: []string{"backupregion1", "backupregion2"},
+				},
+			},
+		},
+	}
+
+	// mock GetOSEnv for testing
+	oldEnvFunc := common.GetOSEnv
+	oldExecFunc := common.StandardExec
+	oldDefaultCred := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+	tf, _ := ioutil.TempFile("", "foo")
+	tf.WriteString(`{"type": "service_account"}`)
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", tf.Name())
+	defer func() {
+		// restore
+		common.GetOSEnv = oldEnvFunc
+		common.StandardExec = oldExecFunc
+		os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", oldDefaultCred)
+		os.Remove(tf.Name())
+	}()
+	// mock as kubectl not set and gcloud set as "b", so check environment
+	// return project as "b"
+	common.StandardExec = func(name string, args ...string) ([]byte, error) {
+		var out []byte
+		var err error
+		switch name {
+		case "gcloud":
+			out = []byte("b")
+			err = nil
+		case "kubectl":
+			out = []byte("")
+			err = fmt.Errorf("kubectl not set")
+		default:
+			out, err = oldExecFunc(name)
+		}
+		return out, err
+	}
+
+	for _, data := range datas {
+		common.GetOSEnv = func(s string) string {
+			switch s {
+			case "E2E_CLUSTER_REGION":
+				return data.regionEnv
+			case "E2E_CLUSTER_BACKUP_REGIONS":
+				return data.backupRegionEnv
+			}
+			return oldEnvFunc(s)
+		}
+		c := GKEClient{}
+		co := c.Setup(data.numNodes, data.nodeType, data.region, data.zone, data.project)
+		errPrefix := fmt.Sprintf("testing setup with:\n\tnumNodes: %v\n\tnodeType: %v\n\tregion: %v\n\tone: %v\n\tproject: %v\n\tregionEnv: %v\n\tbackupRegionEnv: %v",
+			data.numNodes, data.nodeType, data.region, data.zone, data.project, data.regionEnv, data.backupRegionEnv)
+		gotCo := co.(*GKECluster)
+		// mock for easier comparison
+		gotCo.operations = nil
+		if !reflect.DeepEqual(co, data.expClusterOperations) {
+			t.Fatalf("%s\nwant GKECluster:\n'%v'\ngot GKECluster:\n'%v'", errPrefix, data.expClusterOperations, co)
+		}
+	}
+}
+
+func TestInitialize(t *testing.T) {
+	customProj := "customproj"
+	datas := []struct {
```

Comment:
This test should also exercise the Prow case.

Reply:
This test would actually acquire a Boskos project in Prow, I have updated to mock it always return false for IsProw(), so that there is no more accident of acquiring Boskos projects before my next update. Will explore how to mock boskos for this test